### PR TITLE
AGS: NEON and SSE2 blending and blitting optimizations

### PR DESCRIPTION
**Pull request by wyatt-radkiewicz**
Wrote better and more optimized blending/blitting functions for ags's
allegro library. These are targeted mainly towards cpus with SSE2 or NEON,
but optimizations have also been made to the old code that was for any cpu.

### CODE REVIEW

**Comment by tag2015** [comment]
Hi, may I ask which games you tested the SSE2 optimizations with? I built your PR on Win10 with an intel CPU and
found some issues (mostly segfaults, but also a couple problems with wrong colors)

**Comment by wyatt-radkiewicz** [comment]
> Hi, may I ask which games you tested the SSE2 optimizations with? I built your PR on Win10 with an intel CPU and
> found some issues (mostly segfaults, but also a couple problems with wrong colors)

I tested it so far on Kings Quest 2 AGDI edition and Murder Cases Demo with Win10 and an intel CPU but haven't gotten any weird colors or segfaults. I'll fix this before GSOC, but currently I'm working on a different part of my GSOC project.

**Comment by tag2015** [comment]
> I tested it so far on Kings Quest 2 AGDI edition and Murder Cases Demo with Win10 and an intel CPU but haven't gotten any weird colors or segfaults. I'll fix this before GSOC, but currently I'm working on a different part of my GSOC project.

Ok, when you have time or taking a break from the other task it'd be nice if you could test something else such as the following demos:
https://store.steampowered.com/app/1346360/Old_Skies/
https://www.moddb.com/games/the-cat-lady/downloads/the-cat-lady-demo-free
http://www.screen7.co.uk/games/download/Downfall2016_Demo_1.0_Setup.exe
https://bestgames.itch.io/rock-n-roll-will-never-die

File: engines/ags/engine/main/engine.cpp [hunk 1]
```diff
@@ -23,6 +23,7 @@
 // Engine initialization
 //
 
+#include "ags/lib/allegro/color.h"
```
Comment (sev-) [inline engines/ags/engine/main/engine.cpp:26]: I think, this is a leftover from some early experiments

File: engines/ags/globals.cpp [hunk 2]
```diff
@@ -103,10 +103,52 @@ namespace AGS3 {
...
+#else
+	return false;
+#endif
+}
+
 Globals::Globals() {
 	g_globals = this;
 
 	// Allegro globals
+	__bitmap_simd_optimizations = checkForSIMDExtensions();
```
Comment (sev-) [inline engines/ags/globals.cpp:151]: Why two underscores at the beginning?

File: engines/ags/lib/allegro/surface_simd_neon.cpp [hunk 3]
```diff
@@ -0,0 +1,496 @@
+#include "ags/lib/allegro/surface_simd_neon.h"
```
Comment (sev-) [inline engines/ags/lib/allegro/surface_simd_neon.cpp:1]: The file is missing ScummVM copyright header

File: engines/ags/lib/allegro/surface_simd_none.cpp [hunk 4]
```diff
@@ -0,0 +1,36 @@
+#include "ags/lib/allegro/surface.h"
```
Comment (sev-) [inline engines/ags/lib/allegro/surface_simd_none.cpp:1]: Please add the copyright header

File: engines/ags/lib/allegro/surface_simd_ppc.cpp [hunk 5]
```diff
@@ -0,0 +1,512 @@
+// TODO: Complete PowerPC code
```
Comment (sev-) [inline engines/ags/lib/allegro/surface_simd_ppc.cpp:1]: Do we need this file at this moment?

File: engines/ags/lib/allegro/surface_simd_neon.cpp [hunk 6]
```diff
@@ -0,0 +1,496 @@
...
+
+#ifdef AGS_LIB_ALLEGRO_SURFACE_SIMD_NEON_IMPL
+
+#include "ags/lib/allegro/gfx.h"
+#include "ags/lib/allegro/color.h"
+#include "ags/lib/allegro/flood.h"
+#include "ags/ags.h"
+#include "ags/globals.h"
+#include "common/textconsole.h"
```
Comment (sev-) [inline engines/ags/lib/allegro/surface_simd_neon.cpp:10]: Please move common/ and graphics/ includes before ags/ ones

File: engines/ags/lib/allegro/surface_simd_ppc.h [hunk 7]
```diff
@@ -0,0 +1,473 @@
+/* ScummVM - Graphic Adventure Engine
```
Comment (sev-) [inline engines/ags/lib/allegro/surface_simd_ppc.h:1]: Same thing. I think this file could be removed

File: engines/ags/lib/allegro/surface_simd_sse.cpp [hunk 8]
```diff
@@ -0,0 +1,515 @@
+#include "ags/lib/allegro/surface_simd_sse.h"
```
Comment (sev-) [inline engines/ags/lib/allegro/surface_simd_sse.cpp:1]: Copyright header...

File: engines/ags/lib/allegro/surface_simd_sse.cpp [hunk 9]
```diff
@@ -0,0 +1,515 @@
...
+#ifdef AGS_LIB_ALLEGRO_SURFACE_SIMD_SSE_IMPL
+
+#include "ags/lib/allegro/gfx.h"
+#include "ags/lib/allegro/color.h"
+#include "ags/lib/allegro/flood.h"
+#include "ags/ags.h"
+#include "ags/globals.h"
+#include "common/textconsole.h"
```
Comment (sev-) [inline engines/ags/lib/allegro/surface_simd_sse.cpp:9]: includes order...

File: engines/ags/lib/allegro/surface_simd_sse.cpp [hunk 10]
```diff
@@ -0,0 +1,515 @@
...
+	return _mm_cvtsi128_si32(_mm_shuffle_epi32(x, _MM_SHUFFLE(3, 3, 3, 3)));
+}
+
+// This template handles 2bpp and 4bpp, the other specializations handle 1bpp and format conversion blits
+template<int DestBytesPerPixel, int SrcBytesPerPixel, int ScaleThreshold>
+void BITMAP::drawInner4BppWithConv(int yStart, int xStart, uint32 transColor, uint32 alphaMask, PALETTE palette, int useTint, int sameFormat, const ::Graphics::ManagedSurface &src, ::Graphics::Surface &destArea, int horizFlip, int vertFlip, int skipTrans, int srcAlpha, int tintRed, int tintGreen, int tintBlue, const Common::Rect &dstRect, const Common::Rect &srcArea, const BlenderMode blenderMode, int scaleX, int scaleY) {
+	const int xDir = horizFlip ? -1 : 1;
+	byte rSrc, gSrc, bSrc, aSrc;
+	byte rDest = 0, gDest = 0, bDest = 0, aDest = 0;
+    __m128i tint = _mm_sll_epi32(_mm_set1_epi32(srcAlpha), _mm_set1_epi32(24));
```
Comment (sev-) [inline engines/ags/lib/allegro/surface_simd_sse.cpp:33]: Spaces sneaked in into indentation, and on line 39

File: engines/ags/lib/allegro/surface_simd_sse.h [hunk 11]
```diff
@@ -0,0 +1,469 @@
...
+	diffs[1] = _mm_and_si128(_mm_add_epi16(diffs[1], destComps[1]), _mm_set1_epi16(0x3f));
+	diffs[2] = _mm_and_si128(_mm_add_epi16(diffs[2], destComps[2]), _mm_set1_epi16(0x1f));
+
+	// We compile all the colors into diffs[0] as a 16 bit rgb pixel
+	diffs[0] = _mm_or_si128(diffs[0], _mm_slli_epi16(diffs[1], 5));
+	return _mm_or_si128(diffs[0], _mm_slli_epi16(diffs[2], 11));
+}
+
+inline __m128i mul32_as32(__m128i a, __m128i b)
+{
```
Comment (sev-) [inline engines/ags/lib/allegro/surface_simd_sse.h:107]: Code formatting

File: engines/ags/lib/std/functional.h [hunk 12]
```diff
@@ -49,7 +49,7 @@ struct function {
 		return *_fn;
 	}
 
-	operator bool() const {
```
Comment (sev-) [inline engines/ags/lib/std/functional.h:52]: Why is this removed? This method does not alter the class, thus, `const`

File: engines/ags/tests/test_gfx.cpp [hunk 13]
```diff
@@ -19,28 +19,257 @@
...
+												__m128i dest = _mm_set1_epi16((destB >> 3) | ((destG >> 2) << 5) | ((destR >> 3) << 11));
+												__m128i alphas = _mm_set1_epi16((uint16)alpha);
+												simd2bppCol = (uint16)(_mm_cvtsi128_si32(blendPixelSIMD2Bpp(src, dest, alphas)) & 0xffff);
+#endif
+											}
+#ifdef VERBOSE_TEST_GFX
+											debug("src argb: %d, %d, %d, %d dest argb: %d, %d, %d, %d a: %d", srcA, srcR, srcG, srcB, destA, destR, destG, destB, alpha);
+#endif
+											switch ((BlenderMode)blenderMode) {
+												case kSourceAlphaBlender: debug("blenderMode: kSourceAlphaBlender"); break;
```
Comment (sev-) [inline engines/ags/tests/test_gfx.cpp:184]: ... we indent switch() statement differently. Please check our Code Formatting Conventions

**Review by sev- — CHANGES_REQUESTED** [review]
Outstanding work!

I have only a few minuscule notes. Then it could be merged

**Comment by sev-** [comment]
Thank you!

**Pull request merged**